torch/csrc/LazyTensor.cpp

```diff
@@ -0,0 +1,385 @@
+#ifndef C10_MOBILE
+
+#include <ATen/core/LazyTensor.h>
+#include <torch/csrc/jit/graph_executor.h>
+#include <torch/csrc/jit/interpreter.h>
+#include <torch/csrc/jit/operator.h>
+#include <torch/csrc/jit/script/function_schema_parser.h>
+
+
+/* LazyTensor Design
+ *
+ *
+ * LazyTensor is an abstraction that enables lazy execution in PyTorch.  It does so by
+ * providing a virtual tensor implementation that looks and feels like a regular Tensor
+ * but does not necessarily execute any of the underlying computation immediately.
+ *
+ * Instead of executing the computation "eagerly,"  LazyTensors will record requested
+ * operations into an implicit graph and later lower that graph to TorchScript IR.
+ * Once `to_eager()` is called, it is guaranteed that any computation required to
+ * resolve the tensor has been completed.
+ *
+ * For backends that digest TorchScript IR, this design is interesting because it enables
+ * these backends to be used in PyTorch without the need for a TorchScript frontend.
+ *
+ * Note: shape propagation is not supported! This means `to_eager` is required before
+ * any shape dependent code is run.  The implication is that some core modules,
+ * such as nn.BatchNorm, cannot be executed lazily because their implementations
+ * check shapes in Python.
+ *
+ * Note: LazyTensor also accumulates a graph during calls to `backward`.  Shape
+ * information is required for error checking during backward propagation,
+ * but with LazyTensors we skip these checks.
+ *
+ * Note: LazyTensor assumes all operations are re-computable!  This assumption
+ * saves us from using too much memory without explicit need.  However, it also
+ * induces potentially unwanted re-compute and BREAKS semantics for operations
+ * that require state.  TODO: blacklist broken ops
+ *
+ * Note: LazyTensors cannot be used in multi-threaded environment.  This paradigm
+ * should subvert the need for parallelism at this level anyway.
```

Comment:
Reapplying my earlier comment here. LazyTensors is not thread safe. But is this a *fundamental* limitation of the way it is designed, or just an accident / way to make the first implementation easier? If I think of lazy tensor as solely a system of recording an implicit graph, I don't see why it would not be thread safe; indeed, the natural implementation of this ought to be thread safe automatically!

Even when I treat lazy tensors as a form of lazy evaluation ala Haskell, it should *still* be thread safe; section 8 of https://simonmar.github.io/bib/papers/multicore-ghc.pdf gives a good example of how GHC implements lazy evaluation in a thread safe manner. Our performance requirements are much less stringent here, and a simple version of this scheme ought to work.

But I am all guessing here, because all this comment says is that "LazyTensors cannot be used in a multi-threaded environment".